openmeter/ent/schema/subscription.go

```diff
@@ -0,0 +1,135 @@
+package schema
+
+import (
+	"entgo.io/ent"
+	"entgo.io/ent/dialect"
+	"entgo.io/ent/schema/edge"
+	"entgo.io/ent/schema/field"
+	"entgo.io/ent/schema/index"
+
+	"github.com/openmeterio/openmeter/openmeter/productcatalog/plan"
+	"github.com/openmeterio/openmeter/pkg/currencyx"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+)
+
+type Subscription struct {
+	ent.Schema
+}
+
+func (Subscription) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+		entutils.CadencedMixin{},
+	}
+}
+
+func (Subscription) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("plan_key").NotEmpty().Immutable(),
+		field.Int("plan_version").Min(1).Immutable(),
+		field.String("customer_id").NotEmpty().Immutable(),
+		field.String("currency").GoType(currencyx.Code("")).MinLen(3).MaxLen(3).NotEmpty().Immutable(),
+	}
+}
+
+func (Subscription) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "customer_id"),
+	}
+}
+
+func (Subscription) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("prices", Price.Type),
+		edge.To("entitlements", SubscriptionEntitlement.Type),
+		edge.From("customer", Customer.Type).Field("customer_id").Ref("subscription").Immutable().Unique().Required(),
+		edge.To("phases", SubscriptionPhase.Type),
+	}
+}
+
+type SubscriptionPhase struct {
+	ent.Schema
+}
+
+func (SubscriptionPhase) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (SubscriptionPhase) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("subscription_id").NotEmpty().Immutable(),
+		field.String("key").NotEmpty().Immutable(),
+		field.String("name").NotEmpty(),
+		field.String("description").Optional().Nillable(),
+		field.Time("active_from").Immutable(),
+	}
+}
+
+func (SubscriptionPhase) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "subscription_id"),
+		index.Fields("namespace", "subscription_id", "key"),
+	}
+}
+
+func (SubscriptionPhase) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("subscription", Subscription.Type).Field("subscription_id").Ref("phases").Unique().Immutable().Required(),
+		edge.To("items", SubscriptionItem.Type),
+	}
+}
+
+type SubscriptionItem struct {
+	ent.Schema
+}
+
+func (SubscriptionItem) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+		entutils.CadencedMixin{},
+	}
+}
+
+func (SubscriptionItem) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("phase_id").NotEmpty().Immutable(),
+		field.String("key").NotEmpty().Immutable(),
```

Comment:
Do we need a key for it, isn't ID enaough? Do we expect users to name subscription items manually?

Reply:
To avoid linking I'll copy from relevant docs

SubscriptionItem can be identified, it has a reference:
- If a Feature is associated with the SubscriptionItem, it is identified by the Feature
  - It can be an ID reference, for an exact version of the Feature (Features can change across versions)
  - It can be a Key reference, which always refers to the latest (active or inactive) version of a Feature
- If a Feature is not associated with the SubscriptionItem, it is referenced by the Price
  - We say “referenced by the Price” regardless of how a price itself is referenced, it colloquially makes sense to say “paying the same price for the same thing”
In practice this will need an ID _(key)_ being generated

Keys are used for identifying the "same thing" across different instances (with different IDs)